python/dcr_template.py

```diff
@@ -19,137 +20,156 @@
-                            if _j > y0+dy: continue
-                        model_vals = self._extract_model_vals(_j, _i, radius=pix_radius, fft=self.use_fft)
-                        template_vals = self._apply_dcr_kernel(dcr_kernel, model_vals)
-                        self._insert_template_vals(_j, _i, template_vals, template, weights,
-                                                   radius=pix_radius)
-                template[weights > 0] /= weights[weights > 0]
-                template[weights == 0] = 0.0
-                if add_noise:
-                    variance_level = np.median(exp.getMaskedImage().getVariance().getArray())
-                    rand_gen = np.random
-                    template += rand_gen.normal(scale=np.sqrt(variance_level), size=template.shape)
-
-            dataId_out = dataId[_img]
+                dcr_kernel = self._calc_offset_phase(**make_kernel_kwargs)
+
+            template = np.zeros((self.y_size, self.x_size))
+            weights = np.zeros((self.y_size, self.x_size))
+            pix_radius = self.kernel_size//2
+            for j in range(self.y_size):
+                for i in range(self.x_size):
+                    if self._edge_test(j, i):
+                        continue
+                    model_vals = self._extract_model_vals(j, i, radius=pix_radius,
+                                                          model=self.model, weights=self.weights)
+                    template_vals = self._apply_dcr_kernel(dcr_kernel, model_vals)
+                    self._insert_template_vals(j, i, template_vals, template=template, weights=weights,
+                                               radius=pix_radius, kernel=self.psf_avg)
+            template[weights > 0] /= weights[weights > 0]
+            template[weights == 0] = 0.0
+            if add_noise:
+                variance_level = np.median(calexp.getMaskedImage().getVariance().getArray())
+                rand_gen = np.random
+                template += rand_gen.normal(scale=np.sqrt(variance_level), size=template.shape)
+
+            dataId_out = dataId_out_arr[exp_i]
             exposure = self._create_exposure(template, variance=np.abs(template), snap=0,
                                              elevation=el, azimuth=az, obsid=dataId_out['visit'])
             if output_repository is not None:
                 butler_out.put(exposure, "calexp", dataId=dataId_out)
-            yield(exposure)
+            yield exposure
+
+    @staticmethod
+    def _fetch_metadata(exposure, property_name, default_value=None):
+        """Simple wrapper to extract metadata from an exposure, with some error handling.
+
+        @param exposure: An LSST exposure object
+        @param property_name: String, name of the property to be extracted
+        @param default_value: Value to be returned if the property is not found in the exposure metadata.
+        @return Returns the value of property_name from the metadata of exposure.
```

Comment:
I've changed it to return `None` instead of raising an exception, and also pass in only the meatdata from the exposure.